main/src/ca/uwaterloo/flix/api/lsp/provider/Visitor.scala

```diff
@@ -0,0 +1,850 @@
+    * expression containing a specific variable is collected in a list, via
+    * mutation a variable.
+    *
+    * Note that if a node is not accepted, none of the nodes in it's
+    * subtree will be visited either. For instance, if a `def` is not
+    * accepted, none of the expressions in the body will be visited eiter,
+    * even if they would otherwise be accepted by the acceptor.
+    *
+    * @param root      The AST root node.
+    * @param consumer  A consumer which defines what to do when visiting different types of AST nodes.
+    * @param acceptor  An acceptor which defines the the criteria for whether an AST node should be visited.
+    */
+  def visitRoot(root: Root, consumer: Consumer, acceptor: Acceptor): Unit = {
+
+    implicit val c = consumer
+    implicit val a = acceptor
+
+    root.defs.foreach{ case (_, defn) => visitDef(defn) }
+
+    root.effects.foreach{ case (_, eff) => visitEffect(eff) }
+
+    // root.entryPoint.map{ case v => visitEntryPoint(visit, accept)(v) }
+
+    root.enums.foreach{ case (_, e) => visitEnum(e) }
+
+    root.instances.foreach{ case (_, l) => l.foreach(visitInstance) }
+
+    // root.modules.map { case (_, v) => visitModule(v, ???, ???) }
+
+    root.sigs.foreach{ case (_, sig) => visitSig(sig) }
+
+    root.structs.foreach{ case (_, struct) => visitStruct(struct) }
+
+    // root.traitEnv.map{ case (_, v) => visitTraitEnv(???, ???)(v) };
+
+    root.traits.foreach{ case (_, traitt) => visitTrait(traitt) }
+
+    root.typeAliases.foreach{ case (_, alias) => visitTypeAlias(alias) }
+
+    // root.uses
+  }
+
+  def inside(uri: String, pos: Position)(loc: SourceLocation): Boolean = {
+    val (x, y) = pos.toZeroIndexed
+
+    val posLine = x + 1
+    val posCol = y + 1
+
+    // sp1 and sp2, by invariant, has the same source, so we can use either
+    (uri == loc.sp1.source.name) &&
+    (posLine >= loc.beginLine) &&
+    (posLine <= loc.endLine) &&
+    (!(posLine == loc.beginLine && posCol < loc.beginCol)) &&
+    (!(posLine == loc.endLine && posCol >= loc.endCol)) // geq because end column is exclusive
+  }
+
+  def inside(loc1: SourceLocation, loc2: SourceLocation): Boolean = {
+    loc1.source == loc2.source &&
+    (loc1.beginLine >= loc2.beginLine) &&
+    (loc1.endLine <= loc2.endLine) &&
+    !(loc2.beginLine == loc1.beginLine && loc2.beginCol > loc1.beginCol) &&
+    !(loc1.endLine == loc2.endLine && loc1.endCol > loc2.endCol)
+  }
+
+  private def visitEnum(enm: Enum)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(enm.loc)) { return }
+
+    c.consumeEnum(enm)
+
+    visitAnnotations(enm.ann)
+    // TODO visit modifiers?
+    enm.tparams.foreach(visitTypeParam)
+    visitDeriveList(enm.derives)
+    enm.cases.map(_._2).foreach(visitCase)
+  }
+
+  private def visitDeriveList(deriveList: Derivations)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(deriveList.loc)) { return }
```

Comment:
You had something nice going where there was always a line break after the check. Stick to it!